File: src/infrastructure/messaging/kafka_consumer.py
import asyncio
import json
from aiokafka import AIOKafkaConsumer
from aiokafka.errors import KafkaError
import structlog
from src.config.settings import BROKER_URL
from src.domain.models import Evento

logger = structlog.get_logger()

async def start_kafka_consumer():
    try:
        consumer = AIOKafkaConsumer(
            'event_topic',
            bootstrap_servers=BROKER_URL,
            group_id="flow-logic-group",
            auto_offset_reset="earliest"
        )
        await consumer.start()
        logger.info("Kafka consumer connected successfully")

        try:
            async for message in consumer:
                event_data = json.loads(message.value.decode('utf-8'))
                evento = Evento(**event_data)
                logger.info("Received event", event=event_data)
                # Process the event
        finally:
            await consumer.stop()
    except KafkaError as e:
        logger.error("Kafka connection error", error=str(e))

async def main():
    await start_kafka_consumer()

if __name__ == "__main__":
    asyncio.run(main())

File: src/infrastructure/messaging/kafkaConsumer.ts
import { Kafka, Consumer, ConsumerRunConfig, ConsumerConfig, Message } from 'kafkajs';
import { config } from '../../config';
import logger from '../../config/logger';
import { flujoOrchestrator } from '../../application/flujoOrchestrator';

// Create a Kafka instance
const kafka = new Kafka({
  clientId: 'flow-logic-service',
  brokers: [config.BROKER_URL],
});

// Create a consumer instance
const consumer = kafka.consumer({ groupId: 'flow-logic-service-group' });

// Function to connect the consumer
export async function connectConsumer() {
  try {
    await consumer.connect();
    logger.info('Kafka consumer connected successfully');
  } catch (error) {
    logger.error('Error connecting to Kafka consumer:', error);
    process.exit(1);
  }
}

// Function to disconnect the consumer
export async function disconnectConsumer() {
  await consumer.disconnect();
  logger.info('Kafka consumer disconnected');
}

// Function to subscribe to topics and process messages
export async function subscribeToTopics(topics: string[]) {
  await consumer.subscribe({ topics, fromBeginning: true });

  await consumer.run({
    eachMessage: async ({ topic, partition, message }) => {
      try {
        logger.info(
          `Received message: ${message.value?.toString() || ''} ` +
          `from topic: ${topic} partition: ${partition}`
        );

        // Extract event data from Kafka message
        const eventData = message.value ? JSON.parse(message.value.toString()) : {};

        // Create event object
        const event = {
          type: eventData.type,
          payload: eventData.payload,
          metadata: eventData.metadata
        };

        // Record initial flow state if using audit DB
        // await flowStateManager.recordReceived(event.id);

        // Process event through orchestrator
        await flujoOrchestrator.processEvent(event, eventData.flowId);

        logger.info(`Event ${event.type} processed successfully`);
      } catch (error) {
        logger.error(`Error processing message from topic ${topic}:`, error);
      }
    },
  } as ConsumerRunConfig);
}

// Export the consumer for use in other modules
export default consumer;